Source/Scene/ModelExperimental/CustomShaderStage.js

```diff
@@ -0,0 +1,519 @@
+    );
+  }
+
+  if (defined(customShader.fragmentShaderText)) {
+    shaderBuilder.addDefine(
+      "HAS_CUSTOM_FRAGMENT_SHADER",
+      undefined,
+      ShaderDestination.FRAGMENT
+    );
+
+    // add defines like CUSTOM_SHADER_MODIFY_MATERIAL
+    var shaderModeDefine = CustomShaderMode.getDefineName(customShader.mode);
+    shaderBuilder.addDefine(
+      shaderModeDefine,
+      undefined,
+      ShaderDestination.FRAGMENT
+    );
+  }
+
+  var uniforms = customShader.uniforms;
+  for (var uniformName in uniforms) {
+    if (uniforms.hasOwnProperty(uniformName)) {
+      var uniform = uniforms[uniformName];
+      shaderBuilder.addUniform(uniform.type, uniformName);
+    }
+  }
+
+  var varyings = customShader.varyings;
+  for (var varyingName in varyings) {
+    if (varyings.hasOwnProperty(varyingName)) {
+      var varyingType = varyings[varyingName];
+      shaderBuilder.addVarying(varyingType, varyingName);
+    }
+  }
+
+  // if present, the lighting model overrides the material's lighting model.
+  if (defined(customShader.lightingModel)) {
+    renderResources.lightingOptions.lightingModel = customShader.lightingModel;
+  }
+
+  var alphaOptions = renderResources.alphaOptions;
+  if (customShader.isTranslucent) {
+    alphaOptions.pass = Pass.TRANSLUCENT;
+    alphaOptions.alphaMode = AlphaMode.BLEND;
+  } else {
+    // Use the default pass (either OPAQUE or 3D_TILES), regardless of whether
+    // the material pipeline stage used translucent. The default is configured
+    // in AlphaPipelineStage
+    alphaOptions.pass = undefined;
+    alphaOptions.alphaMode = AlphaMode.OPAQUE;
+  }
+
+  renderResources.uniformMap = combine(
+    renderResources.uniformMap,
+    customShader.uniformMap
+  );
+};
+
+function getAttributeNames(attributes) {
+  var names = {};
+  for (var i = 0; i < attributes.length; i++) {
+    var attribute = attributes[i];
+    var semantic = attribute.semantic;
+    var setIndex = attribute.setIndex;
+
+    var variableName;
+    if (defined(semantic)) {
+      variableName = VertexAttributeSemantic.getVariableName(
+        semantic,
+        setIndex
+      );
+    } else {
+      // Handle user defined vertex attributes. They must begin with an underscore
+      // For example, "_TEMPERATURE" will be converted to "a_temperature".
+      variableName = attribute.name.substring(1).toLowerCase();
+    }
+
+    names[variableName] = attribute;
+  }
+  return names;
+}
+
+function generateStructField(glslType, fieldName) {
+  return "    " + glslType + " " + fieldName + ";";
+}
+
+function generateAttributeField(name, attribute) {
+  var attributeType = attribute.type;
+  var glslType = AttributeType.getGlslType(attributeType);
+
+  // Fields for the Attribute struct. for example:
+  // "    vec3 position;"
+  return generateStructField(glslType, name);
+}
+
+// GLSL types of standard attribute types when uniquely defined
+var attributeTypeLUT = {
+  position: "vec3",
+  normal: "vec3",
+  tangent: "vec4",
```

Comment:
Yes, it would appear as a vec4. I haven't worked with tangents in shaders before, what would you expect to see here? it seems that glTF allows for different orientations via the w channel, is that something typically needed even for other types of 3D models?

Okay you're the second one to bring this up. The thing though is I don't have a VertexAttributeSemantic, I have a string from the shader here. So I'd still need a LUT from "normal" -> "NORMAL" -> "vec3" here... what do we want to do?

Reply:
Let's discuss a bit offline.

Reply:
We decided that this is a larger issue beyond what we have time for this week. I'll write up an issue.

Reply:
Opened https://github.com/CesiumGS/cesium/issues/9771 which refers to this thread.